src/cargo/core/compiler/mod.rs

```diff
@@ -393,6 +393,22 @@ fn link_targets<'a, 'cfg>(
         .collect();
     let json_messages = bcx.build_config.json_messages();
 
+    // Find the executable for the current target (if any).
+    let mut executable : Option<PathBuf> = Option::None;
+    for output in cx.outputs(unit)?.iter() {
+        if output.flavor == FileFlavor::DebugInfo {
+            continue;
+        }
+        let bindst = match output.hardlink {
+            Some(ref link_dst) => link_dst,
+            None => &output.path,
+        };
+        if unit.target.is_bin() || unit.target.is_bin_example() {
```

Comment:
I think the logic here needs to be adjusted slightly. In Cargo, there are several flavors of excutables:

1) binaries and examples, where the user writes `fn main()` themselves 
2) tests and becnhmars, where `fn main()` is synthesized by the compiler
3) build scripts

I think we definitely should specify "executable" for 1 and 2, and maybe for 3 as well. 

The test setup for it would look like this:

```
λ exa -GT
.
├── benches
│  └── benchmark.rs
├── Cargo.toml
├── src
│  ├── lib.rs
│  └── main.rs
└── tests
   └── integration_test.rs
```

Then, executing `cargo test --no-run` should show executables for `integration_test` and for `lib.rs` (when testing, libary is compiled as an executable which launches unit-tests).

Executing `cargo bench --no-run` should show exetable for `benchmark`.

Reply:
@matklad I'm having some problems with the tests here. It seems likte that the json lines emitted by `cargo test` isn't emitted in a deterministic order. Am I missing something here or might this be a bug?

Reply:
I think this is ok: for example, Cargo could compile tests and becnhmarks in parallel, so either can be ready first. You could add `with_json_unordered` method to the testing infra to avoid the problem. It also should be possible to modify the test such that there's only one executable per compilation. For example, you can run `cargo build --lib`, `cargo build --bin` separately.

Reply:
@matklad Ah, didn't know about the `with_json_unordered` approach :+1: I'm going with separate assert passes like you suggest since that one sounds cleaner somehow. Thanks!